### PR TITLE
Renode improvements

### PR DESCRIPTION
**Pull request by Citrullin**
### Contribution description

Some improvements for the renode integration. Especially in context of the [RIOT OS IDE](https://github.com/Citrullin/RIOT-OS-IDE). Eventually the IDE contains everything someone needs to start with developing an application in RIOT.
No extras needed. In order to make this possible, I need to change the renode implementation. 
Overall this is still a work in progress. 
Known issues:
- [No such command or device: next_value](https://github.com/renode/renode/issues/152)
- `telnet localhost 3456` seems not to work, even though the UART should be exposed via the [socket terminal](https://renode.readthedocs.io/en/latest/host-integration/uart.html#socket-terminal) on the nrf52840dk. 

### Testing procedure

With RIOT OS snap package: 

1. Clone the [RIOT OS IDE](https://github.com/Citrullin/RIOT-OS-IDE). 
2. Change into the cloned directory. Create a snap with `snapcraft`
3. Install the generated snap with `snap install --devmode riot-os-ide_0.1_[ARCHITECTURE].snap`
4. Change into an example application in RIOT
5. Execute `RENODE=/snap/bin/riot-os-ide.renode BOARD=nrf52840dk make emulate`

With global renode installation:
1. Change into an example application in RIOT
2. Execute `RENODE=/snap/bin/riot-os-ide.renode BOARD=nrf52840dk make emulate`

### Issues/PRs references
Contains workarounds for [#144 in renode](https://github.com/renode/renode/issues/144). 

@basilfx Does this workaround work for you too? Or do I need to change something?
@chrysn Since you also commit on [EFM2RIOT](https://github.com/basilfx/EFM2RIOT/tree/feature/simplify), you are might be interested in this as well.

### CODE REVIEW

**Comment by basilfx** [comment]
Hey @Citrullin, thanks for PR! I was also working on some Renode improvements, which I have in [here](https://github.com/basilfx/RIOT/tree/feature/renode).

I haven't had the chance to test this out, since I don't have a Linux desktop distro at hand right now.

Could you try to squash some commits and adhere to the `area: description` convention? I have the feeling there are some commits that don't belong here ;-) Let me know if you need some help.

File: Makefile.include [hunk 1]
```diff
@@ -730,7 +730,7 @@ debug-server:
 	$(call check_cmd,$(DEBUGSERVER),Debug server program)
 	$(DEBUGSERVER) $(DEBUGSERVER_FLAGS)
 
-emulate:
+emulate: all
```
Comment (basilfx) [inline Makefile.include:733]: Out of curiosity: why is this needed?

File: boards/cc2538dk/dist/board.resc [hunk 2]
```diff
@@ -1,30 +1,35 @@
-mach create
-using sysbus
-machine LoadPlatformDescription @platforms/cpus/cc2538.repl
+:name: CC2538
+:description: This script runs Contiki-NG hello-world example on CC2538.
+
+$name?="CC2538"
+$bin?=@https://dl.antmicro.com/projects/renode/cc2538-contiki_hello_world.elf-s_242120-08fc83d11f790ccc1aa46abfdfc9c2e1a94baed2
```
Comment (basilfx) [inline boards/cc2538dk/dist/board.resc:5]: Check my version at https://github.com/basilfx/RIOT/tree/feature/renode.

File: boards/ikea-tradfri/Makefile.features [hunk 3]
```diff
@@ -10,5 +10,8 @@ FEATURES_PROVIDED += periph_spi
 FEATURES_PROVIDED += periph_timer
 FEATURES_PROVIDED += periph_uart periph_uart_modecfg
 
+# Various other features (if any)
+FEATURES_PROVIDED += emulator_renode
```
Comment (basilfx) [inline boards/ikea-tradfri/Makefile.features:14]: I see this at multiple places, but it doesn't seem to be used, right?

File: dist/tools/renode/run-renode.sh [hunk 4]
```diff
@@ -11,41 +11,90 @@
...
+# Default path to Renode cpu intermediate configuration file
+: ${RENODE_CPU_BIN_REPL:=${BINDIR}/cpu.repl}
 
-# Default path to Renode configuration file
-: ${RENODE_CONFIG:=${BOARDSDIR}/${BOARD}/dist/board.resc}
-# Default path to Renode intermediate configuration file
-: ${RENODE_BIN_CONFIG:=${BINDIR}/board.resc}
 # Default Renode command
-: ${RENODE:=renode}
+if [ ! -f "${RENODE}" ]; then
```
Comment (basilfx) [inline dist/tools/renode/run-renode.sh:46]: Isn't `${RENODE:=renode}` doing the same?

**Review by basilfx — COMMENTED** [review]
A few comments here and there.

**Comment by Citrullin** [comment]
> 
> 
> Hey @Citrullin, thanks for PR! I was also working on some Renode improvements, which I have in [here](https://github.com/basilfx/RIOT/tree/feature/renode).
> 
> I haven't had the chance to test this out, since I don't have a Linux desktop distro at hand right now.
> 
> Could you try to squash some commits and adhere to the `area: description` convention? I have the feeling there are some commits that don't belong here ;-) Let me know if you need some help.

Yes, will do. I've seen Renode already fixed the issue. That was quite fast. Didn't expect that.
You are already using the variables in the renode configuration. So, it would make sense to just fork your solution and adapt my current code to it. What do you think? Are you already done or still working on it?

File: Makefile.include (hunk 1 above)
Comment (Citrullin) [reply Makefile.include:733]: When you only call `make emulate` the binary doesn't get compiled. So I added `all` so that the binary gets compiled. Otherwise it fails when you haven't called another target before.

File: boards/ikea-tradfri/Makefile.features (hunk 3 above)
Comment (Citrullin) [reply boards/ikea-tradfri/Makefile.features:14]: I checked out your code. I don't know. I guess not.

**Comment by basilfx** [comment]
> Are you already done or still working on it?

I don't think I need to change much to 'improve' the Renode integration. My main goal was to have the EFM32 working again. There are still some issues in the emulation code, but that is out of scope of the PR. 

I haven't decided yet if I want to move all CPU and board files (*.repl) files to `dist/tools/renode`, or keep them in `board/<name>/dist/renode` and `cpu/<name>/dist/renode`. What do you think?

File: dist/tools/renode/run-renode.sh (hunk 4 above)
Comment (Citrullin) [reply dist/tools/renode/run-renode.sh:46]: I guess so. I am not that good in shell scripting. Haven't seen that before.

File: Makefile.include (hunk 1 above)
Comment (basilfx) [reply Makefile.include:733]: That sounds like a good idea :-) I'll cherry-pick that commit in my branch.

Comment (basilfx) [reply Makefile.include:733]: It's in: https://github.com/basilfx/RIOT/commit/77473003c60df767bfb7d7b57f5c319a5c21c9ee

**Comment by Citrullin** [comment]
> 
> 
> > Are you already done or still working on it?
> 
> I don't think I need to change much to 'improve' the Renode integration. My main goal was to have the EFM32 working again. There are still some issues in the emulation code, but that is out of scope of the PR.
> 
> I haven't decided yet if I want to move all CPU and board files (*.repl) files to `dist/tools/renode`, or keep them in `board/<name>/dist/renode` and `cpu/<name>/dist/renode`. What do you think?

That is also an issue I ran into. Still not completely sure. If you take a look into `boards/stm32f4discovery/dist/board.repl`, the pattern `board` and `cpu` pattern has its limitations. It fits better into the RIOT structure though. It also makes more sense to have everything for the board at one place and for the cpu at one place. My problem is just that RIOTs stm32_common code doesn't fit into the model of Renodes. I guess `board/<name>/dist/renode` and `cpu/<name>/dist/renode` makese more sense, even though that means more work in restructuring the .repl files.

File: boards/ikea-tradfri/Makefile.features (hunk 3 above)
Comment (basilfx) [reply boards/ikea-tradfri/Makefile.features:14]: Sorry for 'blaming' you, I actually found out I wrote that myself in #8153 :-P
